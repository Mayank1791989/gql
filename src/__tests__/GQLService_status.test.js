/* @flow */
/* eslint-disable max-len */
import runGQLService from './runGQLService';

describe('Schema', () => {
  it('should report errors in schema', (done) => {
    const gql = runGQLService(
      {
        '/test/.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
        '/test/schema/schema.gql': `
          type Query {
            viewer: xViewer # missing viewer
          }
      `,
      },
      {
        cwd: '/test',
        onInit() {
          expect(gql.status()).toMatchSnapshot();
          done();
        },
      },
    );
  });

  it('can modify validation rule severity', (done) => {
    const gql = runGQLService(
      {
        '/test/schema/schema.gql': `
          type Query {
            name: String
          }
          type Hello {
            name: String
          }
        `,
        '/test/.gqlconfig': `{
          schema: {
            files: 'schema/*.gql',
            validate: {
              extends: 'gql-rules-schema',
              rules: {
                NoUnusedTypeDefinition: 'error',
              },
            }
          }
        }`,
      },
      {
        cwd: '/test',
        onInit: () => {
          expect(gql.status()).toMatchSnapshot();
          done();
        },
      },
    );
  });

  it('calling status before onInit should not throw', () => {
    const gql = runGQLService(
      {
        '/test/.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
        '/test/schema/schema.gql': `
          type Query {
            viewer: xViewer # missing viewer
          }
      `,
      },
      {
        cwd: '/test',
      },
    );
    expect(() => gql.status()).not.toThrow();
    expect(gql.status()).toEqual([]);
  });

  it('can turn off validation rules', (done) => {
    const gql = runGQLService(
      {
        '/test/schema/schema.gql': `
          type Query {
            name: String
          }
          type Hello {
            name: String
          }
        `,
        '/test/.gqlconfig': `{
          schema: {
            files: 'schema/*.gql',
            validate: {
              extends: 'gql-rules-schema',
              rules: {
                NoUnusedTypeDefinition: 'off',
              },
            }
          }
        }`,
      },
      {
        cwd: '/test',
        onInit: () => {
          expect(gql.status()).toMatchSnapshot();
          done();
        },
      },
    );
  });
});

describe('Query', () => {
  it('report query errors', (done) => {
    const gql = runGQLService(
      {
        '/test/.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            },
            query: {
              files: [
                {
                  match: 'query/*.gql',
                  parser: 'QueryParser',
                }
              ]
            }
          }
        `,
        '/test/schema/schema.gql': `
          type Query {
            viewer: Viewer
          }
          type Viewer {
            name: String
          }
        `,
        '/test/query/query.gql': `
          query {
            xViewer {
              name
            }
          }
        `,
      },
      {
        cwd: '/test',
        onInit: () => {
          expect(gql.status()).toMatchSnapshot();
          done();
        },
      },
    );
  });
});
