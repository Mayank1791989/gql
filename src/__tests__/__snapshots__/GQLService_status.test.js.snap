// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query report query errors 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 13,
        "line": 3,
        "path": "/test/query/query.gql",
      },
    ],
    "message": "Cannot query field \\"xViewer\\" on type \\"Query\\". Did you mean \\"viewer\\"? (FieldsOnCorrectType)",
    "severity": "error",
  },
]
`;

exports[`Schema can modify validation rule severity 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 11,
        "line": 5,
        "path": "/test/schema/schema.gql",
      },
    ],
    "message": "Unused type definition 'Hello' (NoUnusedTypeDefinition)",
    "severity": "error",
  },
]
`;

exports[`Schema can turn off validation rules 1`] = `Array []`;

exports[`Schema should report errors in schema 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 21,
        "line": 3,
        "path": "/test/schema/schema.gql",
      },
    ],
    "message": "Type \\"xViewer\\" not found.",
    "severity": "error",
  },
]
`;
